### PR TITLE
Implement CefGeolocationHandler::OnRequestGeolocationPermission

### PR DESCRIPTION
**Pull request by pushplay**
delegating the request to a CefSharp.IGeolocationHandler implementation.  This allows CefSharp to permit geolocation requests.

This implements https://github.com/cefsharp/CefSharp/issues/332 .

### CODE REVIEW

**Comment by pushplay** [comment]
This allows for geolocation requests with the option {enableHighAccuracy: false} to work as expected.  A call to the google api domain is made which fails if the google api key isn't set.  There's information about setting the google api key here: http://www.chromium.org/developers/how-tos/api-keys .  I can add a wiki entry on the subject once (if?) the PR is accepted.

I haven't yet confirmed {enableHighAccuracy: true} working as expected on a device with GPS as I don't have one available to test with.  I might be able to get one in the future.

**Comment by amaitland** [comment]
> I can add a wiki entry on the subject once (if?) the PR is accepted.

@pushplay Looks great, let's merge this before the `39` release :+1: A `wiki` entry would be great :+1:

File: CefSharp.Core/Internals/ClientAdapter.cpp [hunk 1]
```diff
@@ -569,5 +569,31 @@ namespace CefSharp
...
+        bool ClientAdapter::OnRequestGeolocationPermission(CefRefPtr<CefBrowser> browser, const CefString& requesting_url, int request_id, CefRefPtr<CefGeolocationCallback> callback)
+        {
+            IGeolocationHandler^ handler = _browserControl->GeolocationHandler;
+            if (handler == nullptr)
+            {
+                // Default deny, as CEF does.
+                return false;
+            }
+
+            if (handler->OnRequestGeolocationPermission(_browserControl, StringUtils::ToClr(requesting_url), request_id)) {
```
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Brace on new line please :+1:

Comment (pushplay) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: @amaitland done :+1:

**Comment by amaitland** [comment]
Thanks, will take it for a spin when I get a chance :+1:

**Comment by jornh** [comment]
@pushplay excellent looking (first?) PR, :+1: really happy to have people doing stuff like that in the community :smile:

**Comment by pushplay** [comment]
It was totally selfish.  I needed it to work, so I did it.  :)

**Comment by jornh** [comment]
:smile: I like that kind of selfishness

**Comment by jankurianski** [comment]
I was able to get this going in so far as seeing my CefSharp geolocation API requests showing as 4xx client errors in the Google dev console... looks like I have to "contact sales" to get actual quota for that API :smile: 

- I think you could add this as a commented-out part of `CefExample.Init()` to show how to set the keys:
```C#
// Set Google API keys, used for Geolocation requests.  See http://www.chromium.org/developers/how-tos/api-keys
// Environment.SetEnvironmentVariable("GOOGLE_API_KEY", "");
// Environment.SetEnvironmentVariable("GOOGLE_DEFAULT_CLIENT_ID", "");
// Environment.SetEnvironmentVariable("GOOGLE_DEFAULT_CLIENT_SECRET", "");
```

- The example `GeolocationHandler` permission message could have a reference to setting API keys in `CefExample.Init()`, that way it is self-documenting.  For example,
> {0} wants to use your computer's location.  Allow?
> ** You must set your Google API key in CefExample.Init() for this to work. **

- Is there any reason to have 2 separate copies of `GeolocationHandler.cs` instead of 1 in the `CefSharp.Example` project?

**Comment by jankurianski** [comment]
BTW, really good PR! :+1:

**Comment by amaitland** [comment]
> I think you could add this as a commented-out part of CefExample.Init() to show how to set the keys:

Nice touch, I like it :+1: 

> Is there any reason to have 2 separate copies of GeolocationHandler.cs instead of 1 in the CefSharp.Example project?

Can you use the `WinForms` messagebox from a `WPF` app? Can't say that I've ever tried it.

**Comment by pushplay** [comment]
> Can you use the WinForms messagebox from a WPF app?

I haven't used UI elements from both in a project, but I have had references to both in a project and it leads to some confusing Intellisense results.

**Comment by jankurianski** [comment]
Sorry, I misread the WPF GeolocationHandler as using the WinForms MessageBox already. Disregard that point :speak_no_evil: :smile:

**Comment by amaitland** [comment]
@pushplay Do you mind making the first two changes @jankurianski suggested? After that I think this can be merged :+1:

**Comment by pushplay** [comment]
@amaitland Sure I'll look into it tomorrow.

**Comment by amaitland** [comment]
@pushplay Fantastic thanks! :+1: 

@jankurianski When your happy feel free to test out that shiny new merge button :smile:

**Comment by jankurianski** [comment]
Thanks a lot for your contribution @pushplay :+1: 
Also added your feature to the [ChangeLog](https://github.com/cefsharp/CefSharp/wiki/ChangeLog) for the 39 release.

**Comment by amaitland** [comment]
@pushplay Greatly appreciated :smile: 

@jankurianski excellent suggestions :+1:

**Pull request merged**